<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Métodos-de-pandas" data-toc-modified-id="Métodos-de-pandas-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Métodos de pandas</a></span><ul class="toc-item"><li><span><a href="#Métodos-para-explorar-nuestro-DataFrame" data-toc-modified-id="Métodos-para-explorar-nuestro-DataFrame-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Métodos para explorar nuestro <em>DataFrame</em></a></span></li><li><span><a href="#Métodos-para-modificar-algunas-características" data-toc-modified-id="Métodos-para-modificar-algunas-características-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Métodos para modificar algunas características</a></span></li></ul></li><li><span><a href="#Reasignación-columnas-df[‘x’]-=-df[‘x’]" data-toc-modified-id="Reasignación-columnas-df[‘x’]-=-df[‘x’]-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Reasignación columnas df[‘x’] = df[‘x’]</a></span></li></ul></div>

**Como siempre antes de empezar tenemos que hacer el import!!**

In [1]:
import pandas as pd

# Métodos de pandas 

En este apartado vamos a ver algunos de los principales que podemos aplicar a nuestros *DataFrames*. El principal objetivo de estos métodos es hacer una exploración inicial de nuestro DataFrame, aunque algunos de ellos nos servirán para hacer algunos cambios en nuestro DataFrame, como por ejemplo cambiar el nombre de las columnas o reemplazar ciertos caracteres. Algunos de los métodos son: 

- Métodos para exploración básica del *DataFrame*: 

    - `.head()`
    - `.tail()`
    - `.sample()`
    - `.columns`
    - `.shape`
    - `.describe()`
    - `.info()`
    - `.unique`
    - `.value_counts()`
    - `.isnull()`
    - `.isna()`
    - `.corr()`

- Métodos para modificar algunas características básicas de nuestro *DataFrame*: 

    - `.set_index()`
    - `.reset_index()`
    - `.rename()`
    - `.drop()`
    - `.cut()`
    - `.replace()`
    

📌Antes de empezar vamos a cargar un dataset para poder trabajar. En Python tenemos una serie de librerias que nos van a proveer de datasets para que nosotros podamos trabajar. En este caso usaremos la librería `PyDataset`

- Lo primero que tenemos que hacer es instalarla desde la terminal

  `pip install pydataset`

  o 

  `pip3 install pydataset`

- Una vez instalado lo importamos en nuestro jupyter: 

  ```python
  from pydataset import data
  ```

In [2]:
from pydataset import data

initiated datasets repo at: /home/pfuente/.pydataset/


In [6]:
# para saber que datasets tenemos 
dfs = data()
dfs

dataset_id                                             title
0    AirPassengers       Monthly Airline Passenger Numbers 1949-1960
1          BJsales                 Sales Data with Leading Indicator
2              BOD                         Biochemical Oxygen Demand
3     Formaldehyde                     Determination of Formaldehyde
4     HairEyeColor         Hair and Eye Color of Statistics Students
..             ...                                               ...
752        VerbAgg                  Verbal Aggression item responses
753           cake                 Breakage Angle of Chocolate Cakes
754           cbpp                 Contagious bovine pleuropneumonia
755    grouseticks  Data on red grouse ticks from Elston et al. 2001
756     sleepstudy       Reaction times in a sleep deprivation study

[757 rows x 2 columns]

Si nos fijamos tenemos 757 datasets, donde tenemos información sobre su título y un `id` el cual usaremos para cargar ese dataset. En este jupyter usaremos un dataset con información sobre plantas petrolíferas.

In [4]:
# cargamos el DataFrame con el que queremos trabajar

df = data('rock')

In [7]:
df.head()

area     peri     shape  perm
1  4990  2791.90  0.090330   6.3
2  7002  3892.60  0.148622   6.3
3  7558  3930.66  0.183312   6.3
4  7352  3869.32  0.117063   6.3
5  7943  3948.54  0.122417  17.1

## Métodos para explorar nuestro *DataFrame*

- `head()`

Nos permite ver las primeras lineas de nuestro *DataFrame*. Por defecto, nos devolverá las 5 primeras lineas. 

In [8]:
df.head()

area     peri     shape  perm
1  4990  2791.90  0.090330   6.3
2  7002  3892.60  0.148622   6.3
3  7558  3930.66  0.183312   6.3
4  7352  3869.32  0.117063   6.3
5  7943  3948.54  0.122417  17.1

In [7]:
# en caso de que queramos más filas solo tendremos que especificar el número entre paréntesis

df.head(7)

area     peri     shape  perm
1  4990  2791.90  0.090330   6.3
2  7002  3892.60  0.148622   6.3
3  7558  3930.66  0.183312   6.3
4  7352  3869.32  0.117063   6.3
5  7943  3948.54  0.122417  17.1
6  7979  4010.15  0.167045  17.1
7  9333  4345.75  0.189651  17.1

- `tail()` 

Nos permite ver las últimas filas de nuestro *DataFrame*, por defecto 5. Pero al igual que en `head()`, podemos especificar el número de filas que querremos ver. 

In [9]:
df.tail()

area      peri     shape    perm
44  8793  2280.490  0.420477  1300.0
45  3475  1174.110  0.200744   580.0
46  1651   597.808  0.262651   580.0
47  5514  1455.880  0.182453   580.0
48  9718  1485.580  0.200447   580.0

In [10]:
# si queremos ver las últimas 7 lineas 

df.tail(7)

area      peri     shape    perm
42  1016   308.642  0.230081  1300.0
43  5605  1145.690  0.464125  1300.0
44  8793  2280.490  0.420477  1300.0
45  3475  1174.110  0.200744   580.0
46  1651   597.808  0.262651   580.0
47  5514  1455.880  0.182453   580.0
48  9718  1485.580  0.200447   580.0

- `sample()`

Nos muestra filas aleatorias de nuestro *DataFrame*, por defecto nos devuelve solo una, pero podemos especificar el número de filas que queremos ver. 

In [10]:
# por defecto nos devuelve una

df.sample()

area     peri     shape   perm
27  7416  1916.24  0.191802  740.0

In [11]:
# ¿y si queremos ver 5? Lo especificamos entre paréntesis. 

df.sample(5)

area     peri     shape   perm
32   7894  1461.06  0.276016  890.0
39   3524  1189.46  0.163586  100.0
7    9333  4345.75  0.189651   17.1
19  11878  4864.22  0.200071   58.6
8    8209  4344.75  0.164127   17.1

- `shape`

Nos devuelve el número de filas y columnas que tenemos en nuestro *DataFrame*.

In [12]:
# nuestro dataset tiene 48 filas y 4 columnas

df.shape

(48, 4)

- `dtypes`

Nos devuelve el tipo de datos que tenemos en cada una de nuestras columnas.

In [13]:
# en este caso tenemos integers y floats

df.dtypes

area       int64
peri     float64
shape    float64
perm     float64
dtype: object

- `columns`

Nos muestra el nombre de las columnas de nuestro *DataFrame* .

In [14]:
df.columns

Index(['area', 'peri', 'shape', 'perm'], dtype='object')

- `describe()`

Nos devuelve un *DataFrame* con un resumen de los principales estadísticos (media, mediana, desviacón estándar etc.) de nuestras **columnas numéricas**.

In [15]:
df.describe()

area         peri      shape         perm
count     48.000000    48.000000  48.000000    48.000000
mean    7187.729167  2682.211938   0.218110   415.450000
std     2683.848862  1431.661164   0.083496   437.818226
min     1016.000000   308.642000   0.090330     6.300000
25%     5305.250000  1414.907500   0.162262    76.450000
50%     7487.000000  2536.195000   0.198862   130.500000
75%     8869.500000  3989.522500   0.262670   777.500000
max    12212.000000  4864.220000   0.464125  1300.000000

Si nos fijamos esto nos devuelve solo los principales estadísticos de las variables numéricas, pero ... ¿qué pasa con las variables categóricas? Para esto tenemos el parámetro `include` dentro del `describe` donde podremos especificar que tipo de dato queremos. 

En este caso, todas nuestras variables son de tipo numérico, pero podría darse el caso que tengamos variables de tipo *string*, que en Pandas nos aparacerá como tipo **object**. Para poder acceder a los principales estadísticos de estas columnas haremos lo siguiente: 

```python
df.describe(include = "object")
```

- `info()` 

Nos devuelve un pequeño resumen sobre: 

- El número de columnas que tenemos.
- Los nombres de las columnas.
- El número de valores no nulos.
- El tipo de los datos de nuestras columnas.

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48 entries, 1 to 48
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   area    48 non-null     int64  
 1   peri    48 non-null     float64
 2   shape   48 non-null     float64
 3   perm    48 non-null     float64
dtypes: float64(3), int64(1)
memory usage: 1.9 KB


- `unique()`

En este caso tendremos que especificar la columna sobre la que queremos trabajar. 

¿Cómo hacemos esto en pandas? 

```python
# el nombre del DataFrame seguido de corchetes con el nombre de la columna
df["nombre_columna"]

# esto es lo mismo que: 

df.nombre_columna
```

In [17]:
# en la columna area tenemos los siguientes valores únicos. 

df["area"].unique()

array([ 4990,  7002,  7558,  7352,  7943,  7979,  9333,  8209,  8393,
        6425,  9364,  8624, 10651,  8868,  9417,  8874, 10962, 10743,
       11878,  9867,  7838, 11876, 12212,  8233,  6360,  4193,  7416,
        5246,  6509,  4895,  6775,  7894,  5980,  5318,  7392,  3469,
        1468,  3524,  5267,  5048,  1016,  5605,  8793,  3475,  1651,
        5514,  9718])

In [18]:
# ¿y que valores únicos tenemos en la columna perm?

df["perm"].unique()

array([   6.3,   17.1,  119. ,   82.4,   58.6,  142. ,  740. ,  890. ,
        950. ,  100. , 1300. ,  580. ])

- `value_counts()`

Devuelve una serie que contiene el recuento de valores únicos.

El objeto resultante estará en orden descendente para que el primer elemento sea el más frecuente. Excluye los valores NA por defecto.

De nuevo tendremos que especificar que columna es sobre la que queremos aplicar el método. 

In [19]:
# en este caso, para cada uno de nuestros valores únicos tenemos 4 valores en nuestro DataFrame

df["perm"].value_counts()

6.3       4
17.1      4
119.0     4
82.4      4
58.6      4
142.0     4
740.0     4
890.0     4
950.0     4
100.0     4
1300.0    4
580.0     4
Name: perm, dtype: int64

In [19]:
df["peri"].value_counts()

1461.060    2
2791.900    1
1376.700    1
1916.240    1
1585.420    1
1851.210    1
1239.660    1
1728.140    1
1426.760    1
990.388     1
1350.760    1
476.322     1
1977.390    1
1189.460    1
1644.960    1
941.543     1
308.642     1
1145.690    1
2280.490    1
1174.110    1
597.808     1
1455.880    1
1379.350    1
3518.440    1
3892.600    1
3986.240    1
3930.660    1
3869.320    1
3948.540    1
4010.150    1
4345.750    1
4344.750    1
3682.040    1
3098.650    1
4480.050    1
4036.540    1
4697.650    1
3518.040    1
3999.370    1
3629.070    1
4608.660    1
4787.620    1
4864.220    1
4479.410    1
3428.740    1
4353.140    1
1485.580    1
Name: peri, dtype: int64

- `isnull()`

Nos devuelve un booleano con los valores que faltan en nuestro DataFrame.

In [20]:
df.isnull().head()

area   peri  shape   perm
1  False  False  False  False
2  False  False  False  False
3  False  False  False  False
4  False  False  False  False
5  False  False  False  False

Pero si nos fijamos esto es algo difícil de interpretar, ¿cómo solucionamos esto? 

In [21]:
# de esta forma nos devuelve el número de valores nulos que tenemos por columnas. En este caso 0 por columna

df.isnull().sum()

area     0
peri     0
shape    0
perm     0
dtype: int64

- `isna()`

Hace lo mismo que el método anterior, nos devuelve si nuestros valores son nulos o no. 

In [22]:
df.isna().sum()

area     0
peri     0
shape    0
perm     0
dtype: int64

- `corr()`

Calcula la correlación por pares de las columnas, excluyendo los valores NA/nulos. 

Aunque ya lo veremos más adelante, la correlación trata de establecer la relación o dependencia que existe entre dos variables. Cuanto más cerca del 1 esté este valor, más correlacionadas estarán las variables. El signo indica si esa correlación es positiva o negativa. Es decir:

- Positiva: a mayor valor de una variable, mayor valor de la otra. 
- Negativa: a mayor valor de una variable, menor de la otra. 

In [24]:
df.corr()

area      peri     shape      perm
area   1.000000  0.822506 -0.182161 -0.396637
peri   0.822506  1.000000 -0.433126 -0.738716
shape -0.182161 -0.433126  1.000000  0.556721
perm  -0.396637 -0.738716  0.556721  1.000000

## Métodos para modificar algunas características

- `set_index()`

Establezca el índice del DataFrame (etiquetas de fila) utilizando una o más columnas. El índice puede sustituir al índice existente o ampliarlo.

La sintaxis que debemos seguir: 

```python
nuestroDataFrame.set_index(["nombre_columna"], inplace = True)

# con el parámetro `inplace` lo que hacemos es sobreescribir sobre el DataFrame orginal. Es decir, los cambios se ejecutarán sobre nuestro DataFrame sin necesidad de crear uno nuevo
```

In [23]:
df.set_index(["area"], inplace = True)

In [24]:
df.head()

peri     shape  perm
area                         
4990  2791.90  0.090330   6.3
7002  3892.60  0.148622   6.3
7558  3930.66  0.183312   6.3
7352  3869.32  0.117063   6.3
7943  3948.54  0.122417  17.1

In [25]:
# ⚠️ ojo! que cuando ponemos una columna como índice ya no formará parte de nuestras columnas

df.columns

Index(['peri', 'shape', 'perm'], dtype='object')

📌 Si nos fijamos ahora la columna `area` aparece como el índice de nuestro DataFrame. 

- `reset_index()`

De la misma forma que podemos poner una columna como índice, podemos quitarla para que pase a formar parte de las columnas. 

In [26]:
df.reset_index(inplace = True)

In [27]:
df.head()

area     peri     shape  perm
0  4990  2791.90  0.090330   6.3
1  7002  3892.60  0.148622   6.3
2  7558  3930.66  0.183312   6.3
3  7352  3869.32  0.117063   6.3
4  7943  3948.54  0.122417  17.1

- `rename()`

Nos va a permitir modificar el nombre de nuestras columnas. 

📌 Necesitaremos usar diccionarios para usar este método. La sintaxis que aplicaremos será: 

```python
MiDataframe.rename(columns = {"nombre-antiguo": "nombre-nuevo"}, inplace = True)
```

En este caso tenemos los nombres de las columnas en ingles y vamos cambiarlas al español: 

In [28]:
df.rename(columns = {"area": "area", "peri": "perimetro", "shape": "forma", "perm": "perm"}, inplace = True)

In [29]:
# si ahora chequeamos los nombres de las columnas veremos que estos han cambiado. 

df.columns

Index(['area', 'perimetro', 'forma', 'perm'], dtype='object')

Si tenemos muchas columnas que cambiar puede resultar muy tedioso. Pero lo podemos optimizar! 

🤔 ¿Recordais que con las *list comprehensions*? Podremos hacer lo mismo pero creando diccionarios. Vamos a poner en mayúsculas los nombres de las columnas. 

In [30]:
# lo primero que tendremos que hacer es un dict comprehension donde las keys
# serán el nombre de la columnas antiguas y los values el nuevo nombre

columnas_nuevas = {col :  col.upper() for col in df.columns}

In [31]:
# chequeamos si nuestro diccionario está bien

columnas_nuevas

{'area': 'AREA', 'perimetro': 'PERIMETRO', 'forma': 'FORMA', 'perm': 'PERM'}

In [32]:
df.rename(columns = columnas_nuevas, inplace = True)

In [33]:
# Bien 💪 ya tenemos todas nuestras columnas cambiadas!!

df.head()

AREA  PERIMETRO     FORMA  PERM
0  4990    2791.90  0.090330   6.3
1  7002    3892.60  0.148622   6.3
2  7558    3930.66  0.183312   6.3
3  7352    3869.32  0.117063   6.3
4  7943    3948.54  0.122417  17.1

- `drop()`

Elimina las etiquetas especificadas de las filas o columnas.

La sintaxis básica que debemos seguir: 

```python
MiDataframe.drop(labels, axis)
```

Donde: 

- `labels`: Los elementos que queremos eliminar, que además deben ir en una lista. 
- `axis`: hace referencia al eje sobre el que queremos aplicar la eliminación: 
    
    - axis = 1 COLUMNAS
    - axis = 0 FILAS

In [34]:
# En este caso vamos a crear una DataFrame nuevo para almacenar los cambios de eliminar filas o columnas. 
# Como lo haremos con columnas: queremos eliminar la columna "area" de las columnas

df_eliminar = df.drop(["AREA"], axis = 1)

In [35]:
# si nos fijamos hemos eliminado la columna "area" de nuestro DataFrame
df_eliminar.head()

PERIMETRO     FORMA  PERM
0    2791.90  0.090330   6.3
1    3892.60  0.148622   6.3
2    3930.66  0.183312   6.3
3    3869.32  0.117063   6.3
4    3948.54  0.122417  17.1

In [36]:
# podemos incluso eliminar 1 o más columnas

df_eliminar2 = df.drop(["AREA", "FORMA"], axis = 1)
df_eliminar2.head()

PERIMETRO  PERM
0    2791.90   6.3
1    3892.60   6.3
2    3930.66   6.3
3    3869.32   6.3
4    3948.54  17.1

In [37]:
# ¿ y cómo lo haríamos para las filas? 
# En este caso las "etiquetas" serán los valores de los índices de las filas

df_eliminar3 = df.drop([0,1], axis = 0)
df_eliminar3.head()

AREA  PERIMETRO     FORMA  PERM
2  7558    3930.66  0.183312   6.3
3  7352    3869.32  0.117063   6.3
4  7943    3948.54  0.122417  17.1
5  7979    4010.15  0.167045  17.1
6  9333    4345.75  0.189651  17.1

In [38]:
df.PERIMETRO.unique()

array([2791.9  , 3892.6  , 3930.66 , 3869.32 , 3948.54 , 4010.15 ,
       4345.75 , 4344.75 , 3682.04 , 3098.65 , 4480.05 , 3986.24 ,
       4036.54 , 3518.04 , 3999.37 , 3629.07 , 4608.66 , 4787.62 ,
       4864.22 , 4479.41 , 3428.74 , 4353.14 , 4697.65 , 3518.44 ,
       1977.39 , 1379.35 , 1916.24 , 1585.42 , 1851.21 , 1239.66 ,
       1728.14 , 1461.06 , 1426.76 ,  990.388, 1350.76 , 1376.7  ,
        476.322, 1189.46 , 1644.96 ,  941.543,  308.642, 1145.69 ,
       2280.49 , 1174.11 ,  597.808, 1455.88 , 1485.58 ])

- `cut()`

Se utiliza para separar los elementos de un *DataFrame* en diferentes intervalos. La función de corte se utiliza principalmente para realizar análisis estadísticos de datos escalares.  

Aunque tiene múltiples parámetros, la principal sintaxis que debemos seguir es: 

```python
MiDataframe.cut(x, bins)
```
Donde: 

- `x`: es la columna sobre la que queremos aplicar los intervalos
- `bins`: son los intervalos nuevos que queremos definir

En este caso vamos a aplicar este método a la columna `perm`: 

- Primero vamos a ver cuales son sus valores únicos para saber mejor como definir nuestros nuevos *bins*

    ```python
    df.perm.unique()

    # Output: 
    array([   6.3,   17.1,  119. ,   82.4,   58.6,  142. ,  740. ,  890. ,
            950. ,  100. , 1300. ,  580. ])
    ```

- Una vez que tenemos los valores únicos establecemos los nuevos *bins*. Este parámetro lo tenemos que pasar como lista. 

In [41]:
df['bins'] = pd.cut(x=df['PERM'], bins=[0, 100, 200,300, 400,500, 600,
                                          700, 800, 1000, 1100, 1200, 1300])

In [43]:
# basicamente lo que hemos hecho ha sido crear una nueva columna donde tendremos ciertos intervalos. 

df

AREA  PERIMETRO     FORMA    PERM          bins
0    4990   2791.900  0.090330     6.3      (0, 100]
1    7002   3892.600  0.148622     6.3      (0, 100]
2    7558   3930.660  0.183312     6.3      (0, 100]
3    7352   3869.320  0.117063     6.3      (0, 100]
4    7943   3948.540  0.122417    17.1      (0, 100]
5    7979   4010.150  0.167045    17.1      (0, 100]
6    9333   4345.750  0.189651    17.1      (0, 100]
7    8209   4344.750  0.164127    17.1      (0, 100]
8    8393   3682.040  0.203654   119.0    (100, 200]
9    6425   3098.650  0.162394   119.0    (100, 200]
10   9364   4480.050  0.150944   119.0    (100, 200]
11   8624   3986.240  0.148141   119.0    (100, 200]
12  10651   4036.540  0.228595    82.4      (0, 100]
13   8868   3518.040  0.231623    82.4      (0, 100]
14   9417   3999.370  0.172567    82.4      (0, 100]
15   8874   3629.070  0.153481    82.4      (0, 100]
16  10962   4608.660  0.204314    58.6      (0, 100]
17  10743   4787.620  0.262727    58.6      (0, 100]
18  11878   4864.220  0.200071    58.6      (0, 100]
19   9867   4479.410  0.144810    58.6      (0, 100]
20   7838   3428.740  0.113852   142.0    (100, 200]
21  11876   4353.140  0.291029   142.0    (100, 200]
22  12212   4697.650  0.240077   142.0    (100, 200]
23   8233   3518.440  0.161865   142.0    (100, 200]
24   6360   1977.390  0.280887   740.0    (700, 800]
25   4193   1379.350  0.179455   740.0    (700, 800]
26   7416   1916.240  0.191802   740.0    (700, 800]
27   5246   1585.420  0.133083   740.0    (700, 800]
28   6509   1851.210  0.225214   890.0   (800, 1000]
29   4895   1239.660  0.341273   890.0   (800, 1000]
30   6775   1728.140  0.311646   890.0   (800, 1000]
31   7894   1461.060  0.276016   890.0   (800, 1000]
32   5980   1426.760  0.197653   950.0   (800, 1000]
33   5318    990.388  0.326635   950.0   (800, 1000]
34   7392   1350.760  0.154192   950.0   (800, 1000]
35   7894   1461.060  0.276016   950.0   (800, 1000]
36   3469   1376.700  0.176969   100.0      (0, 100]
37   1468    476.322  0.438712   100.0      (0, 100]
38   3524   1189.460  0.163586   100.0      (0, 100]
39   5267   1644.960  0.253832   100.0      (0, 100]
40   5048    941.543  0.328641  1300.0  (1200, 1300]
41   1016    308.642  0.230081  1300.0  (1200, 1300]
42   5605   1145.690  0.464125  1300.0  (1200, 1300]
43   8793   2280.490  0.420477  1300.0  (1200, 1300]
44   3475   1174.110  0.200744   580.0    (500, 600]
45   1651    597.808  0.262651   580.0    (500, 600]
46   5514   1455.880  0.182453   580.0    (500, 600]
47   9718   1485.580  0.200447   580.0    (500, 600]

- `replace()`

Reemplaza los valores especificados por otro valor que nosotros indiquemos.

Los valores del *DataFrame* se sustituyen por otros valores de forma dinámica.

📌 Esto difiere de la actualización con `.loc` o `.iloc`, que requieren que se especifique una ubicación para actualizar con algún valor.

La sintaxis en este caso: 

```python
MiDataframe.replace(to_replace, value, inplace)
```
Donde:

- `to_replace`: es el valor que queremos reemplazar
- `value`: el valor nuevo que queremos incluir
- `inplace`: para que los cambios se ejecuten sobre el *DataFrame* orginal.

En este caso vamos a querer reemplazar los valores `4990` por `area_nueva`

In [44]:
df.replace(to_replace =4990,
                 value ="AREA_NUEVA", inplace = True)

In [45]:
# si nos fijamos donde antes ponía "4990" ahora pone "area_nueva". Este cambio se aplicaría a todo el DataFrame
df.head()

AREA  PERIMETRO     FORMA  PERM      bins
0  AREA_NUEVA    2791.90  0.090330   6.3  (0, 100]
1        7002    3892.60  0.148622   6.3  (0, 100]
2        7558    3930.66  0.183312   6.3  (0, 100]
3        7352    3869.32  0.117063   6.3  (0, 100]
4        7943    3948.54  0.122417  17.1  (0, 100]

In [46]:
# pero también lo podemos aplicar a una columna específica: 

df["PERM"].replace(to_replace =6.3,
                 value ="error", inplace = True)

In [47]:
# aquí tenemos nuestros "error" en los valores específicados. 

df.head()

AREA  PERIMETRO     FORMA   PERM      bins
0  AREA_NUEVA    2791.90  0.090330  error  (0, 100]
1        7002    3892.60  0.148622  error  (0, 100]
2        7558    3930.66  0.183312  error  (0, 100]
3        7352    3869.32  0.117063  error  (0, 100]
4        7943    3948.54  0.122417   17.1  (0, 100]

# Reasignación columnas df[‘x’] = df[‘x’]

Pandas nos va a permitir sobreescribir los valores de una columna. Esto puede ser el resultado de operaciones entre columnas o determinadas transformaciones que queramos hacerle a la columna. 

Veamos un ejemplo: 

El perímetro está en centímetros y lo vamos a querer transformar a metros.

In [51]:
df["PERIMETRO"] = df["PERIMETRO"] / 100

In [52]:
# Si ahora volvemos a ver la columna, veremos que los valores han cambiado
# pero que no se ha creado ninguna nueva columna 

df

AREA  PERIMETRO     FORMA    PERM          bins
0   AREA_NUEVA   0.279190  0.090330   error      (0, 100]
1         7002   0.389260  0.148622   error      (0, 100]
2         7558   0.393066  0.183312   error      (0, 100]
3         7352   0.386932  0.117063   error      (0, 100]
4         7943   0.394854  0.122417    17.1      (0, 100]
5         7979   0.401015  0.167045    17.1      (0, 100]
6         9333   0.434575  0.189651    17.1      (0, 100]
7         8209   0.434475  0.164127    17.1      (0, 100]
8         8393   0.368204  0.203654   119.0    (100, 200]
9         6425   0.309865  0.162394   119.0    (100, 200]
10        9364   0.448005  0.150944   119.0    (100, 200]
11        8624   0.398624  0.148141   119.0    (100, 200]
12       10651   0.403654  0.228595    82.4      (0, 100]
13        8868   0.351804  0.231623    82.4      (0, 100]
14        9417   0.399937  0.172567    82.4      (0, 100]
15        8874   0.362907  0.153481    82.4      (0, 100]
16       10962   0.460866  0.204314    58.6      (0, 100]
17       10743   0.478762  0.262727    58.6      (0, 100]
18       11878   0.486422  0.200071    58.6      (0, 100]
19        9867   0.447941  0.144810    58.6      (0, 100]
20        7838   0.342874  0.113852   142.0    (100, 200]
21       11876   0.435314  0.291029   142.0    (100, 200]
22       12212   0.469765  0.240077   142.0    (100, 200]
23        8233   0.351844  0.161865   142.0    (100, 200]
24        6360   0.197739  0.280887   740.0    (700, 800]
25        4193   0.137935  0.179455   740.0    (700, 800]
26        7416   0.191624  0.191802   740.0    (700, 800]
27        5246   0.158542  0.133083   740.0    (700, 800]
28        6509   0.185121  0.225214   890.0   (800, 1000]
29        4895   0.123966  0.341273   890.0   (800, 1000]
30        6775   0.172814  0.311646   890.0   (800, 1000]
31        7894   0.146106  0.276016   890.0   (800, 1000]
32        5980   0.142676  0.197653   950.0   (800, 1000]
33        5318   0.099039  0.326635   950.0   (800, 1000]
34        7392   0.135076  0.154192   950.0   (800, 1000]
35        7894   0.146106  0.276016   950.0   (800, 1000]
36        3469   0.137670  0.176969   100.0      (0, 100]
37        1468   0.047632  0.438712   100.0      (0, 100]
38        3524   0.118946  0.163586   100.0      (0, 100]
39        5267   0.164496  0.253832   100.0      (0, 100]
40        5048   0.094154  0.328641  1300.0  (1200, 1300]
41        1016   0.030864  0.230081  1300.0  (1200, 1300]
42        5605   0.114569  0.464125  1300.0  (1200, 1300]
43        8793   0.228049  0.420477  1300.0  (1200, 1300]
44        3475   0.117411  0.200744   580.0    (500, 600]
45        1651   0.059781  0.262651   580.0    (500, 600]
46        5514   0.145588  0.182453   580.0    (500, 600]
47        9718   0.148558  0.200447   580.0    (500, 600]

**EJERCICIOS** 

1️⃣ Usa la librería `PyDataset` para descargarte un *DataFrame*. En este caso puedes elegir el que más te guste! 

2️⃣ Haz una exploración del *DataFrame*:

* Muestra las primeras 5 filas
* ¿Cuántas filas y columnas hay en el *DataFrame*?
* ¿Cuántos nulos hay?
* ¿Cuáles son los principales estádisticos de nuestras columnas numéricas?
* ¿Qué correlación existe entre nuestras columnas? 
    
    **BONUS** ¿Podrías hacer una pequeña interpretación de los resultados que nos sale de la correlación?

* Cambia el nombre de las columnas. Dependiendo del *DataFrame* haremos unos cambios u otros. Algunos cambios que podemos hacer son:

    * Poner los nombres todo en minúsculas o mayúsculas
    * Eliminar los espacios en los nombres
        
3️⃣ Guarda el *DataFrame* con los cambios hechos en un archivo `csv`

In [53]:
df1 = data('HairEyeColor')
df1

Hair    Eye     Sex  Freq
1   Black  Brown    Male    32
2   Brown  Brown    Male    53
3     Red  Brown    Male    10
4   Blond  Brown    Male     3
5   Black   Blue    Male    11
6   Brown   Blue    Male    50
7     Red   Blue    Male    10
8   Blond   Blue    Male    30
9   Black  Hazel    Male    10
10  Brown  Hazel    Male    25
11    Red  Hazel    Male     7
12  Blond  Hazel    Male     5
13  Black  Green    Male     3
14  Brown  Green    Male    15
15    Red  Green    Male     7
16  Blond  Green    Male     8
17  Black  Brown  Female    36
18  Brown  Brown  Female    66
19    Red  Brown  Female    16
20  Blond  Brown  Female     4
21  Black   Blue  Female     9
22  Brown   Blue  Female    34
23    Red   Blue  Female     7
24  Blond   Blue  Female    64
25  Black  Hazel  Female     5
26  Brown  Hazel  Female    29
27    Red  Hazel  Female     7
28  Blond  Hazel  Female     5
29  Black  Green  Female     2
30  Brown  Green  Female    14
31    Red  Green  Female     7
32  Blond  Green  Female     8

In [54]:
# Muestra las primeras 5 filas
df1.head()

Hair    Eye   Sex  Freq
1  Black  Brown  Male    32
2  Brown  Brown  Male    53
3    Red  Brown  Male    10
4  Blond  Brown  Male     3
5  Black   Blue  Male    11

In [57]:
#  ¿Cuántas filas y columnas hay en el *DataFrame*?
df1.shape

(32, 4)

In [58]:
#  ¿Cuántos nulos hay?
df.isnull().sum()

AREA         0
PERIMETRO    0
FORMA        0
PERM         0
bins         0
dtype: int64

In [63]:
# ¿Cuáles son los principales estádisticos de nuestras columnas numéricas?
df1.describe()

Freq
count  32.000000
mean   18.500000
std    18.242099
min     2.000000
25%     7.000000
50%    10.000000
75%    29.250000
max    66.000000

In [62]:
# ¿Qué correlación existe entre nuestras columnas? 
df1.corr()

/tmp/ipykernel_1070/2208519770.py:2: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  df1.corr()


Freq
Freq   1.0

In [ ]:
#  **BONUS** ¿Podrías hacer una pequeña interpretación de los resultados que nos sale de la correlación?


In [67]:
# Cambia el nombre de las columnas. Dependiendo del *DataFrame* haremos unos cambios u otros. Algunos cambios que podemos hacer son:
#  * Poner los nombres todo en minúsculas o mayúsculas
#  * Eliminar los espacios en los nombres

columnas_nuevas = {col :  col.upper() for col in df1.columns}
df1.rename(columns = columnas_nuevas, inplace = True)
df1.head()

HAIR    EYE   SEX  FREQ
1  Black  Brown  Male    32
2  Brown  Brown  Male    53
3    Red  Brown  Male    10
4  Blond  Brown  Male     3
5  Black   Blue  Male    11